improver/calibration/reliability_calibration.py

```diff
@@ -590,6 +621,75 @@ def _create_new_bin_coord(probability_bin_coord, upper):
         )
         return new_bin_coord
 
+    def _combine_undersampled_bins(
+        self,
+        observation_count,
+        forecast_probability_sum,
+        forecast_count,
+        probability_bin_coord,
+    ):
+        """
+        Combine bins that are under-sampled i.e. that have a lower forecast
+        count than the minimum_forecast_count, so that information from these
+        poorly-sampled bins can contribute to the calibration. If multiple
+        bins are below the minimum forecast count, the bin closest to
+        meeting the minimum_forecast_count criterion is combined with whichever
+        neighbour has the lowest sample count. A new bin is then created by
+        summing the neighbouring pair of bins. This process is repeated for all
+        bins that are below the minimum forecast count criterion.
+
+        Args:
+            observation_count (numpy.ndarray):
+                Observation count extracted from reliability table.
+            forecast_probability_sum (numpy.ndarray):
+                Forecast probability sum extracted from reliability table.
+            forecast_count (numpy.ndarray):
+                Forecast count extracted from reliability table.
+            probability_bin_coord (iris.coord.DimCoord):
+                Original probability bin coordinate.
+        Returns:
+            Tuple[numpy.ndarray, numpy.ndarray, numpy.ndarray, iris.coord.DimCoord]
+                Tuple containing the updated observation count,
+                forecast probability sum, forecast count and probability bin
+                coordinate.
+
+        """
+        while any(x < self.minimum_forecast_count for x in forecast_count):
```

Comment:
Just thought of one more interesting edge case we may want a unit test for. What happens if all bins are below the minimum count and after you've combined them all they are still below the limit. You could get stuck in an infinite loop? I think we might need an extra condition in the loop to say as long as there are at least two bins in forecast_count.

Reply:
I've added an additional condition on the while loop and a unit test to cover this edge case.